plugins/embark-assistant/matcher.cpp

```diff
@@ -2717,6 +2722,19 @@ namespace embark_assist {
 //  Visible operations
 //=======================================================================================
 
+void embark_assist::matcher::setup() {
+    embark_assist::matcher::state = new(embark_assist::matcher::states);
+    embark_assist::survey::initiate(&state->mlt);
+}
+
+//=======================================================================================
+
+void embark_assist::matcher::shutdown() {
+    delete state;
```

Comment:
I realize there's not a destructor where this could be easily placed, so to prevent accidental double-frees down the road, I might suggest also setting `state = nullptr` here.

Reply:
That is a good idea, something like 
```
void embark_assist::matcher::shutdown() {
    if (state != nullptr)  {
        delete state;
        state = nullptr;
    }
}
```
Will do.

Reply:
You can make a C array of uint8_t (number of elements / 8 rounded up), iterate by 64 (bits) and assign 0 (uint64_t or int64_t) to the address of the first, 8:th, ... byte through casting until don't have 8 bytes left, at which time you can iterate over bytes (or try a quad byte, a double byte, and a byte to squeeze absolutely everything out of it). That's definitely not pretty, though, so I wouldn't do it for the few seconds saved, but instead keep the "original" code (i.e. the one in this PR) and ditch std::fill.

Reply:
@PatrikLundell: That is pretty much, was the hack I mentioned does as well - by manipulating the underlying data structure directly it is able to do a bulk reset. But in both cases the resulting code is not pretty... :)
But have a look at the numbers I generated in the other comment above. I thought about changing vector<bool> to another type but the memory cost would have been significant - but that is no longer true.

Reply:
I didn't read the hack in detail, but it looked like it was using bytes rather than octo-bytes (and I think it would have to if a vector is used rather than an array, as I think a vector isn't guaranteed to be contiguous).

Reply:
Ah I see what you mean now - yes - it operates on bytes not on octo-bytes - but the idea is the same, which is what I meant. Actually I think the hack is absolutely not guaranteed to be portable either (though it works at least with two versions of MSVC) so the contiguousness of the data is probably the least problem if one tries to use this in a code base like dfhack :D

Reply:
Yes, making assumptions of what the internal structure of the vector implementation is would definitely enter risky territory, which is another reason for using an array rather than vector, as I think the array has a precise definition. However, this is an academic discussion at this point.

Reply:
For reference: as of C++03, vector data is required to be stored contiguously, except for `vector<bool>`.